### PR TITLE
Add PchipInterpolator to cupyx.scipy.interpolate

### PR DESCRIPTION
**Pull request by andfoy**
See https://github.com/cupy/cupy/issues/7186

### CODE REVIEW

File: cupyx/scipy/interpolate/_cubic.py [hunk 1]
```diff
@@ -132,3 +138,184 @@ def __init__(self, x, y, dydx, axis=0, extrapolate=None):
...
+        w1 = 2*hk[1:] + hk[:-1]
+        w2 = hk[1:] + 2*hk[:-1]
+
+        # values where division by zero occurs will be excluded
+        # by 'condition' afterwards
+        whmean = (w1 / mk[:-1] + w2 / mk[1:]) / (w1 + w2)
+
+        dk = cupy.zeros_like(y)
+        dk[1:-1][condition] = 0.0
+        dk[1:-1][~condition] = 1.0 / whmean[~condition]
```
Comment (takagi) [inline cupyx/scipy/interpolate/_cubic.py:270]: How about `cp.where` to reduce kernel calls?

**Comment by takagi** [comment]
Record: PchipInterpolator is derived from CubicHermiteSpline, which is in turn derived from PPoly, we can refer its performance comparison to https://github.com/cupy/cupy/pull/7204#issuecomment-1352117803.

File: cupyx/scipy/interpolate/_cubic.py (hunk 1 above)
Comment (takagi) [reply cupyx/scipy/interpolate/_cubic.py:270]: @andfoy kindly ping

Comment (andfoy) [reply cupyx/scipy/interpolate/_cubic.py:270]: Hi @takagi, I'm sorry for not replying! Unfortunately, I'm out of the office (and away from my computer) until Thursday, I'll update the PR once I get back

**Comment by takagi** [comment]
Windows only fails some tests...

**Comment by andfoy** [comment]
@takagi, it is weird that the failures occur on tests different from the ones that are added on this PR. I'll try to just remove the merge commits and do a clean rebase against master, just to check if something went wrong there

**Comment by takagi** [comment]
Thanks, I've kicked the CI again, and will see the result.

**Review by takagi — APPROVED** [review]
LGTM!

**Comment by takagi** [comment]
The CI failure on cuda120 is the same as filed in https://github.com/cupy/cupy/issues/7240.
https://ci.preferred.jp/cupy.linux.cuda120/121699/

**Comment by takagi** [comment]
Thanks, @andfoy!

**Pull request merged**